File: apps/cli/src/regeneration-hints.ts
import type { TargetDirKind } from './types.js';

const HINTS: Partial<Record<TargetDirKind, string>> = {
  'go-global-cache': '`go clean -cache` / rebuild with `go build`',
  'jvm-global-cache': 'Re-download via Gradle/Maven on next build',
  'ide-global-cache': 'Xcode rebuilds DerivedData on next compile',
  'npm-global-cache': '`npm cache clean --force`',
  'pnpm-global-store': '`pnpm store prune`; packages re-fetch on install',
  'yarn-global-cache': '`yarn cache clean`; reinstall dependencies',
  'pip-global-cache': '`pip cache purge`',
  'uv-global-cache': '`uv cache clean`',
  'conda-pkgs-cache': '`conda clean --all -p` (package cache only; does not remove envs)',
  'cargo-registry-cache': '`cargo cache -a`; crates re-download on next build',
  'bun-global-cache': '`bun pm cache rm`; packages re-fetch on install',
  'nuget-global-cache': 'Restore with `dotnet restore` / `nuget restore`',
  'composer-global-cache': '`composer clear-cache`; packages re-fetch on install',
  'vcpkg-installed-cache': 'Reinstall ports with `vcpkg install` for your triplet',
  'conan-global-cache': '`conan remove "*"` -c; packages re-fetch on next build',
  'ccache-global-cache': '`ccache -C` or delete cache dir; objects rebuild on compile',
  'sccache-global-cache': 'Clear sccache store; compiler cache repopulates on build',
  'bazel-disk-cache':
    'Point Bazel `--disk_cache` at a new directory or delete this tree; rebuild repopulates the cache',
};

export function regenerationHintForKind(kind: TargetDirKind): string | undefined {
  return HINTS[kind];
}

export function displayWithRegenerationHint(
  kind: TargetDirKind,
  reasonCodes: readonly string[],
): string {
  const base =
    reasonCodes.length === 0
      ? 'Unspecified'
      : reasonCodes.map((c) => c.toLowerCase().replace(/_/g, ' ')).join(', ');
  const hint = regenerationHintForKind(kind);
  return hint ? `${base}. Regenerate: ${hint}` : base;
}

File: apps/desktop/src-tauri/src/engine/regeneration_hints.rs
use super::types::Kind;

/// How to repopulate after cleanup (shown in candidate detail). Keep commands conservative.
pub fn regeneration_hint_for_kind(kind: &Kind) -> Option<&'static str> {
    match kind {
        Kind::GoGlobalCache => Some("`go clean -cache` / rebuild with `go build`"),
        Kind::JvmGlobalCache => Some("Re-download via Gradle/Maven on next build"),
        Kind::IdeGlobalCache => Some("Xcode rebuilds DerivedData on next compile"),
        Kind::NpmGlobalCache => Some("`npm cache clean --force`"),
        Kind::PnpmGlobalStore => Some("`pnpm store prune`; packages re-fetch on install"),
        Kind::YarnGlobalCache => Some("`yarn cache clean`; reinstall dependencies"),
        Kind::PipGlobalCache => Some("`pip cache purge`"),
        Kind::UvGlobalCache => Some("`uv cache clean`"),
        Kind::CondaPkgsCache => Some("`conda clean --all -p` (package cache only; does not remove envs)"),
        Kind::CargoRegistryCache => Some("`cargo cache -a`; crates re-download on next build"),
        Kind::BunGlobalCache => Some("`bun pm cache rm`; packages re-fetch on install"),
        Kind::NugetGlobalCache => Some("Restore with `dotnet restore` / `nuget restore`"),
        Kind::ComposerGlobalCache => Some("`composer clear-cache`; packages re-fetch on install"),
        Kind::VcpkgInstalledCache => Some("Reinstall ports with `vcpkg install` for your triplet"),
        Kind::ConanGlobalCache => Some("`conan remove \"*\" -c`; packages re-fetch on next build"),
        Kind::CcacheGlobalCache => Some("`ccache -C` or delete cache dir; objects rebuild on compile"),
        Kind::SccacheGlobalCache => Some("Clear sccache store; compiler cache repopulates on build"),
        Kind::BazelDiskCache => Some("Point Bazel `--disk_cache` at a new directory or delete this tree; rebuild repopulates the cache"),
        _ => None,
    }
}

pub fn display_with_regeneration_hint(kind: &Kind, reason_codes: &[String]) -> String {
    let base = if reason_codes.is_empty() {
        "Unspecified".to_string()
    } else {
        reason_codes
            .iter()
            .map(|code| code.to_lowercase().replace('_', " "))
            .collect::<Vec<_>>()
            .join(", ")
    };
    match regeneration_hint_for_kind(kind) {
        Some(hint) => format!("{base}. Regenerate: {hint}"),
        None => base,
    }
}
